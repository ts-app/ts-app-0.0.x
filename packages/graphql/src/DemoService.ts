import { Resolver } from './Resolver'
import { ServiceInfo } from './ServiceInfo'
import { ResolverMiddlewareInput } from './ResolverMiddlewareInput'

export class DemoService {
  private anotherNumber: number = 123

  constructor (public count: number = 0) {
  }

  @Resolver({ paramNames: [ 'name', 'message' ] })
  echo (name: string, message: string) {
    return `Hello, ${name}! Message is "${message}"`
  }

  @Resolver({ type: 'mutation' })
  add ({ howMany = 1 }: { howMany?: number }) {
    this.count += howMany
    return this.count
  }

  @Resolver({ type: 'mutation' })
  async promiseToSubtract ({ howMany = 1 }: { howMany?: number }): Promise<number> {
    return new Promise<number>(resolve => {
      setTimeout(() => {
        this.count -= howMany
        resolve(this.count)
      }, 1)
    })
  }

  justUseAnotherNumber () {
    return this.anotherNumber * 3
  }

  @Resolver()
  throwSomeError () {
    throw new Error('simulated error')
  }

  @Resolver()
  resolverInfo (param: { title: string }, resolverParam?: ResolverMiddlewareInput) {
    const metadata = JSON.stringify(resolverParam![ 'metadata' ], null, 2)
    const user = JSON.stringify(resolverParam![ 'user' ], null, 2)
    const obj = JSON.stringify(resolverParam![ 'obj' ], null, 2)
    const info = JSON.stringify(resolverParam![ 'info' ], null, 2)
    const args = JSON.stringify(resolverParam![ 'args' ], null, 2)
    return {
      title: param.title,
      metadata, user, obj, info, args
    }
  }

  @Resolver()
  resolverInfoWithoutParam (dummy?: any, resolverParam?: ResolverMiddlewareInput) {
    // resolverParam is still the second argument so place a dummy parameter first.

    const metadata = JSON.stringify(resolverParam![ 'metadata' ], null, 2)
    const user = JSON.stringify(resolverParam![ 'user' ], null, 2)
    const obj = JSON.stringify(resolverParam![ 'obj' ], null, 2)
    const info = JSON.stringify(resolverParam![ 'info' ], null, 2)
    const args = JSON.stringify(resolverParam![ 'args' ], null, 2)
    return {
      metadata, user, obj, info, args
    }
  }
}

export class DemoServiceWithServiceInfo implements ServiceInfo {
  info (): { serviceName: string } {
    return {
      serviceName: 'CustomServiceName'
    }
  }

  @Resolver({ paramNames: [ 'name', 'message' ] })
  echo2 (name: string, message: string) {
    return `Hello, ${name}! Message is "${message}" - from echo2`
  }
}
