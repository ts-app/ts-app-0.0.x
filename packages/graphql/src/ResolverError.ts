// https://github.com/Microsoft/TypeScript/issues/13965
export class ResolverError extends Error {
  // tslint:disable-next-line
  __proto__: Error

  constructor (message: string) {
    const trueProto = new.target.prototype
    super(message)

    // Alternatively use Object.setPrototypeOf if you have an ES6 environment.
    this.__proto__ = trueProto
  }
}
