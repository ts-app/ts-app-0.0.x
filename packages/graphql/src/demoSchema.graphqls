type Query {
  echo (name: String!, message: String!): String
  throwSomeError: String
}

type Mutation {
  add (howMany: Int): Int
  promiseToSubtract (howMany: Int): Int
}
