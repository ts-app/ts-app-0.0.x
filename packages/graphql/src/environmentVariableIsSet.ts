import { ResolverMiddlewareInput } from './ResolverMiddlewareInput'

/**
 * This function is used to enable/disable access to GraphQL resolver based on specified environment
 * variable to be set to '1'.
 *
 * @param {string} name Environment variable name
 * @return {(input: ResolverMiddlewareInput) => boolean}
 */
export const environmentVariableIsSet = (name: string) => {
  const authorized = process.env[ name ] === '1'

  return (input: ResolverMiddlewareInput): boolean => authorized
// tslint:disable-next-line
}
