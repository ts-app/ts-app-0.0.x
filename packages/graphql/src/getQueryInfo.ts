import { SchemaInfo } from './SchemaInfo'

export const getQueryInfo = function (schemaInfo: SchemaInfo) {
  const {
    schema, resolvers: {
      Query, Mutation
    }
  } = schemaInfo

  const resolverQueries = Object.keys(Query)
  const typeDefQueries = Object.keys(schema.getQueryType().getFields())
  const queriesWithoutResolver = typeDefQueries.filter(q => !resolverQueries.includes(q))

  const resolverMutations = Object.keys(Mutation)
  const mutationType = schema.getMutationType()
  const typeDefMutations = Object.keys(mutationType ? mutationType.getFields() : [])
  const mutationsWithoutResolvers = typeDefMutations.filter(m => !resolverMutations.includes(m))

  return { typeDefQueries, typeDefMutations, queriesWithoutResolver, mutationsWithoutResolvers }
}
