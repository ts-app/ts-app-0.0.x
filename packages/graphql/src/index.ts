export * from './demoSchemaDefinition'
export * from './DemoService'
export * from './environmentVariableIsSet'
export * from './mergeSchemaDefinitions'
export * from './mergeTypeDefs'
export * from './Resolver'
export * from './ResolverError'
export * from './ResolverMiddlewareInput'
export * from './ResolverService'
export * from './SchemaDefinition'
export * from './SecurityMiddleware'
export * from './serveSchema'
export * from './ServiceInfo'
export * from './standardSchemaDefinition'
export * from './userIdMatchParam'
