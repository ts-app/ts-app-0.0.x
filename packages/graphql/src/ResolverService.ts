import { Middleware, MiddlewareFunction } from '@ts-app/common'
import { ResolverMiddlewareInput } from './ResolverMiddlewareInput'
import { ResolverError } from './ResolverError'
import { ResolverAuth } from './ResolverAuth'

let logError: (message?: any, ...optionalParams: any[]) => void = console.error
let _INSTANCE: ResolverService

/**
 * ResolverService is the core class that enables functions annotated with @Resolver() to work as
 * GraphQL resolvers.
 *
 * It is a singleton class where:
 * - Resolver() decorated functions are registered.
 * - All service classes with @Resolver() decorated functions must be are registered.
 * - Before/Afterwares can be registered to manage GraphQL queries from a centralized function.
 * - Graphql resolver functions are generated based on ResolverService's state.
 *
 */
export class ResolverService {
  private _serviceRegistry: {
    [className: string]: {
      service?: object
      resolvers: {
        [resolverName: string]: {
          descriptor: TypedPropertyDescriptor<Function>
          queryType: 'mutation' | 'query'
          paramNames: string[] | null
          auth: string[] | ResolverAuth
          name: string | null
        }
      }
    }
  }
  private beforewares: MiddlewareFunction<ResolverMiddlewareInput>[] = []
  private afterwares: MiddlewareFunction<ResolverMiddlewareInput>[] = []

  private constructor () {
    this._serviceRegistry = {}
  }

  static getInstance () {
    if (!_INSTANCE) {
      _INSTANCE = new ResolverService()
    }
    return _INSTANCE
  }

  /**
   * Register instance of service to handle resolver functions. There should only be one instance
   * of service class. Otherwise, it does not compute.
   *
   * @param {T} service service being registered
   * @param {string} className service name. Defaults to class name. WARNING: do not use this
   * parameter. It is here due to bad design/assumption. It should be removed.
   * @return {T} service being registered
   */
  registerService<T> (service: T, className = service.constructor.name): T {
    // --- support ServiceInfo
    if (typeof service.constructor.prototype.info === 'function') {
      const info = service.constructor.prototype.info
      className = info()[ 'serviceName' ]
    }

    if (className in this._serviceRegistry) {
      this._serviceRegistry[ className ].service = service
    } else {
      this._serviceRegistry[ className ] = { service, resolvers: {} }
    }

    return service
  }

  registerResolver (className: string, resolverName: string,
                    descriptor: TypedPropertyDescriptor<Function>, queryType: 'mutation' | 'query',
                    paramNames: string[] | null, auth: string[] | ResolverAuth, name: string | null) {
    let clazz = this._serviceRegistry[ className ]
    if (!clazz) {
      clazz = { resolvers: {} }
      this._serviceRegistry[ className ] = clazz
    }
    clazz.resolvers[ resolverName ] = { descriptor, queryType, paramNames, auth, name }
  }

  resetServices () {
    // leave "resolvers" alone. they are only set once (via @Resolver)

    // reset "service"
    Object.keys(this._serviceRegistry).map(className => {
      delete this._serviceRegistry[ className ].service
    })

    // reset before/after wares
    this.beforewares.length = 0
    this.afterwares.length = 0
  }

  /**
   * Returns a function that is binded to the service instance so that "this" works as expected.
   *
   * @param {string} className
   * @param {string} functionName
   * @return {any}
   */
  composeServiceFunction (className: string, functionName: string) {
    const descriptor = this._serviceRegistry[ className ].resolvers[ functionName ].descriptor
    const service = this._serviceRegistry[ className ].service
    const fn = descriptor.get ? descriptor.get() : descriptor.value
    return fn!.bind(service)
  }

  composeResolver (className: string, functionName: string, type: 'mutation' | 'query' = 'query', auth: string[] | ResolverAuth = [], resolverName: string = '') {
    // when resolver executes, it resolves the middleware
    return async (obj: any, args: any, context: any, info: any) => {
      try {
        const user = context.user || null
        const resolverParams = {
          obj, args, context, info,
          metadata: { className, functionName, resolverName, auth, type },
          user
        }
        const middleware = new Middleware<ResolverMiddlewareInput>([
          ...this.beforewares,
          this.createServiceFnMiddleware(className, functionName, resolverParams),
          ...this.afterwares
        ])

        return await middleware.resolve(resolverParams)
      } catch (e) {
        const now = Date.now()
        logError(`[${now}]`, e)
        if (e instanceof ResolverError) {
          return {
            error: e.message
          }
        } else {
          return {
            error: `Error processing resolver function [${className}.${functionName}] [${now}]`
          }
        }
      }
    }
  }

  makeResolvers () {
    const queryResolvers: { [resolverName: string]: any } = {}
    const mutationResolvers: { [resolverName: string]: any } = {}

    Object.keys(this._serviceRegistry).map(className => {
      const { resolvers, service } = this.serviceRegistry[ className ]
      if (service) {
        Object.keys(resolvers).map(key => {
          const { queryType, auth, name } = resolvers[ key ]
          // use custom resolver name if specified, else use decorated function's name
          const resolverName = name ? name : key
          if (queryType === 'mutation') {
            if (resolverName in mutationResolvers) {
              throw new Error(`Mutation resolver [${resolverName}] must be unique`)
            }
            mutationResolvers[ resolverName ] = this.composeResolver(className, key, 'query', auth, resolverName)
          } else {
            if (resolverName in queryResolvers) {
              throw new Error(`Query resolver [${resolverName}] must be unique`)
            }
            queryResolvers[ resolverName ] = this.composeResolver(className, key, 'mutation', auth, resolverName)
          }
        })
      }
    })

    return {
      Query: queryResolvers,
      Mutation: mutationResolvers
    }
  }

  setLogger ({ error }: { error: (message?: any, ...optionalParams: any[]) => void }) {
    logError = error
  }

  getService<T extends Object> (serviceType: Function | string): T | null {
    let serviceName

    if (typeof serviceType === 'string') {
      serviceName = serviceType
    } else if (typeof serviceType.prototype.info === 'function') {
      // --- support ServiceInfo
      serviceName = serviceType.prototype.info()[ 'serviceName' ]
    } else {
      serviceName = serviceType.name
    }

    // const serviceName = typeof serviceType === 'string' ? serviceType : serviceType.name
    const serviceObj = this._serviceRegistry[ serviceName ]
    if (serviceObj && serviceObj.service) {
      return serviceObj.service as T
    } else {
      return null
    }
  }

  registerBeforeware (before: MiddlewareFunction<ResolverMiddlewareInput>) {
    this.beforewares.push(before)
  }

  registerAfterware (after: MiddlewareFunction<ResolverMiddlewareInput>) {
    this.afterwares.push(after)
  }

  private createServiceFnMiddleware (className: string, functionName: string, resolverParams: object): MiddlewareFunction<ResolverMiddlewareInput> {
    // service function that performs business logic
    const serviceFn = this.composeServiceFunction(className, functionName)

    const resolver = this._serviceRegistry[ className ].resolvers[ functionName ]
    const { paramNames } = resolver
    if (paramNames) {
      // middleware transform args from object to array based on paramNames
      return ({ obj, args, context, info }, next) => {
        // convert args (object) to as array, pass to serviceFn as spread parameters
        const params = paramNames.map(name => args[ name ])
        next(serviceFn(...params, resolverParams))
      }
    } else {
      // middleware just pass args to serviceFn (it is expected to take a single object parameter)
      return ({ obj, args, context, info }, next) => {
        next(serviceFn(args, resolverParams))
      }
    }
  }

  get serviceRegistry () {
    return this._serviceRegistry
  }
}
