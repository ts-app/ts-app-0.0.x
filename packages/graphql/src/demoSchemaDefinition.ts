import { loadFile } from '@ts-app/common'
import { DemoService } from './DemoService'
import { ResolverService } from './ResolverService'
import { standardSchemaDefinition } from './standardSchemaDefinition'
import { SchemaDefinition } from './SchemaDefinition'

export const demoSchemaDefinition = (): SchemaDefinition => {
  const resolver = ResolverService.getInstance()
  resolver.registerService(new DemoService())
  const typeDefs = loadFile(`${__dirname}/demoSchema.graphqls`)
  const resolvers = resolver.makeResolvers()
  const standard = standardSchemaDefinition()

  return {
    typeDefs,
    resolvers,
    dependencies: {
      standard
    }
  }
}
