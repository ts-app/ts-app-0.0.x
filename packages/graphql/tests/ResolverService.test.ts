import { ResolverService } from '../src/ResolverService'
import { DemoService, DemoServiceWithServiceInfo } from '../src/DemoService'
import { ResolverError } from '../src/ResolverError'

describe('ResolverService', () => {
  const resolverService = ResolverService.getInstance()
  let demoService: DemoService
  let demoServiceWithServiceInfo: DemoServiceWithServiceInfo

  beforeEach(function () {
    demoService = new DemoService()
    expect(demoService).toBeTruthy()
    resolverService.registerService(demoService)

    demoServiceWithServiceInfo = new DemoServiceWithServiceInfo()
    expect(demoServiceWithServiceInfo).toBeTruthy()
    resolverService.registerService(demoServiceWithServiceInfo)
  })

  afterEach(() => {
    resolverService.resetServices()
    resolverService.setLogger({ error: console.error })
  })

  test('registerService()', () => {
    expect(resolverService.serviceRegistry[ 'DemoService' ].service).toMatchSnapshot()
    // expect(() => resolverService.registerService(new DemoService())).toThrowError()
  })

  test('registerResolver()', () => {
    const clazz = resolverService.serviceRegistry[ 'DemoService' ]
    expect(clazz.resolvers).toMatchSnapshot()
  })

  test('composeServiceFunction()', () => {
    const add = resolverService.composeServiceFunction('DemoService', 'add')
    add({ howMany: 2 })
    add({ howMany: 8 })
    expect(demoService.count).toBe(10)
  })

  test('composeResolver() for sync function', async () => {
    const add = resolverService.composeResolver('DemoService', 'add')
    await add({}, { howMany: 3 }, {}, {})
    await add({}, { howMany: 5 }, {}, {})
    expect(demoService.count).toBe(8)
  })

  test('composeResolver() for async function (Promise)', async () => {
    const promiseToSubtract = resolverService.composeResolver('DemoService', 'promiseToSubtract')
    await promiseToSubtract({}, { howMany: 3 }, {}, {})
    await promiseToSubtract({}, { howMany: 5 }, {}, {})
    expect(demoService.count).toBe(-8)
  })

  test('composeResolver() with paramNames', async () => {
    const echo = resolverService.composeResolver('DemoService', 'echo')
    const output = await echo({}, { name: 'Alice', message: 'Wassup!' }, {}, {})
    expect(output).toMatchSnapshot()
  })

  test('makeResolvers()', () => {
    expect(resolverService.makeResolvers()).toMatchSnapshot()
  })

  test('makeResolvers() - resolvers are usable', async () => {
    const resolvers = resolverService.makeResolvers()

    const queryOutput = await resolvers.Query.echo({}, {
      name: 'Alice',
      message: 'Wassup!'
    }, {}, {})
    expect(queryOutput).toMatchSnapshot()

    const mutationOutput = await resolvers.Mutation.promiseToSubtract({}, { howMany: 10 }, {}, {})
    expect(mutationOutput).toBe(-10) // return value is correct
    expect(demoService.count).toBe(-10) // also affects underlying service state
  })

  test('composeResolver() functions that throw error', async () => {
    let logErrorCount = 0
    resolverService.setLogger({
      error: message => logErrorCount += 1
    })
    const throwSomeError = resolverService.composeResolver('DemoService', 'throwSomeError')
    const output = await throwSomeError({}, {}, {}, {})
    expect(output.error).toMatch(/Error processing resolver function \[DemoService.throwSomeError\]/)
    expect(logErrorCount).toBe(1)
  })

  test('getService() works', () => {
    expect(resolverService.getService(DemoService)).toBe(demoService)
  })

  test('getService() returns null if service not registered', () => {
    class NoInstanceService {
    }

    expect(resolverService.getService(NoInstanceService)).toBeNull()
  })

  test('getService() with ServiceInfo', () => {
    expect(resolverService.getService(DemoServiceWithServiceInfo)).toBe(demoServiceWithServiceInfo)
  })

  test('registerBeforeware() works', async () => {
    let trail: any[] = []

    resolverService.registerBeforeware((input, next) => {
      trail.push(input)
      next('beforeware 1 was called')
    })
    resolverService.registerBeforeware((input, next, prev) => {
      trail.push(prev)
      next('beforeware 2 was called')
    })

    const add = resolverService.composeResolver('DemoService', 'add')
    const three = await add({}, { howMany: 3 }, {}, {})
    expect(demoService.count).toBe(3)
    expect(three).toBe(3)
    expect(trail).toMatchSnapshot()
  })

  test('sample authorization beforeware', async () => {
    resolverService.setLogger({
      error: () => {
        // do nothin
      }
    })

    resolverService.registerBeforeware((input, next) => {
      throw new ResolverError('Unauthorized access')
    })

    const add = resolverService.composeResolver('DemoService', 'add')
    const unauthorized = await add({}, { howMany: 3 }, {}, {})

    // internal state not changed
    expect(demoService.count).toBe(0)
    expect(unauthorized).toEqual({ error: 'Unauthorized access' })
  })

  test('registerAfterware() works', async () => {
    const trail: any[] = []

    resolverService.registerAfterware((input, next, prev) => {
      trail.push('after-1')
      next(prev)
    })
    resolverService.registerAfterware((input, next, prev) => {
      trail.push('after-2')
      next(prev)
    })

    const add = resolverService.composeResolver('DemoService', 'add')
    const three = await add({}, { howMany: 3 }, {}, {})
    expect(demoService.count).toBe(3)
    expect(three).toBe(3)
    expect(trail).toMatchSnapshot()
  })

  test('ServiceInfo support', async () => {
    const service = resolverService.getService('CustomServiceName')
    expect(typeof service).toBe('object')

    const echo2 = resolverService.composeResolver('CustomServiceName', 'echo2')
    const output = await echo2({}, { name: 'Bob', message: 'Up! Up! Up!' }, {}, {})
    expect(output).toMatchSnapshot()
  })
})
