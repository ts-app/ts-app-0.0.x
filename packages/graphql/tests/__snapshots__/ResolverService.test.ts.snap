// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResolverService ServiceInfo support 1`] = `"Hello, Bob! Message is \\"Up! Up! Up!\\" - from echo2"`;

exports[`ResolverService composeResolver() with paramNames 1`] = `"Hello, Alice! Message is \\"Wassup!\\""`;

exports[`ResolverService makeResolvers() - resolvers are usable 1`] = `"Hello, Alice! Message is \\"Wassup!\\""`;

exports[`ResolverService makeResolvers() 1`] = `
Object {
  "Mutation": Object {
    "add": [Function],
    "promiseToSubtract": [Function],
  },
  "Query": Object {
    "echo": [Function],
    "echo2": [Function],
    "resolverInfo": [Function],
    "resolverInfoWithoutParam": [Function],
    "throwSomeError": [Function],
  },
}
`;

exports[`ResolverService registerAfterware() works 1`] = `
Array [
  "after-1",
  "after-2",
]
`;

exports[`ResolverService registerBeforeware() works 1`] = `
Array [
  Object {
    "args": Object {
      "howMany": 3,
    },
    "context": Object {},
    "info": Object {},
    "metadata": Object {
      "auth": Array [],
      "className": "DemoService",
      "functionName": "add",
      "resolverName": "",
      "type": "query",
    },
    "obj": Object {},
    "user": null,
  },
  "beforeware 1 was called",
]
`;

exports[`ResolverService registerResolver() 1`] = `
Object {
  "add": Object {
    "auth": Array [],
    "descriptor": Object {
      "configurable": true,
      "enumerable": true,
      "value": [Function],
      "writable": true,
    },
    "name": null,
    "paramNames": null,
    "queryType": "mutation",
  },
  "echo": Object {
    "auth": Array [],
    "descriptor": Object {
      "configurable": true,
      "enumerable": true,
      "value": [Function],
      "writable": true,
    },
    "name": null,
    "paramNames": Array [
      "name",
      "message",
    ],
    "queryType": "query",
  },
  "promiseToSubtract": Object {
    "auth": Array [],
    "descriptor": Object {
      "configurable": true,
      "enumerable": true,
      "value": [Function],
      "writable": true,
    },
    "name": null,
    "paramNames": null,
    "queryType": "mutation",
  },
  "resolverInfo": Object {
    "auth": Array [],
    "descriptor": Object {
      "configurable": true,
      "enumerable": true,
      "value": [Function],
      "writable": true,
    },
    "name": null,
    "paramNames": null,
    "queryType": "query",
  },
  "resolverInfoWithoutParam": Object {
    "auth": Array [],
    "descriptor": Object {
      "configurable": true,
      "enumerable": true,
      "value": [Function],
      "writable": true,
    },
    "name": null,
    "paramNames": null,
    "queryType": "query",
  },
  "throwSomeError": Object {
    "auth": Array [],
    "descriptor": Object {
      "configurable": true,
      "enumerable": true,
      "value": [Function],
      "writable": true,
    },
    "name": null,
    "paramNames": null,
    "queryType": "query",
  },
}
`;

exports[`ResolverService registerService() 1`] = `
DemoService {
  "anotherNumber": 123,
  "count": 0,
}
`;
